tools/telemetry-generator/src/handlers/stageTimingRetriever.ts

```diff
@@ -10,60 +10,90 @@ interface ParsedJob {
+		throw new Error("No records found in the input data.");
 	}
 	if (process.env.BUILD_ID === undefined) {
-		console.log("BUILD_ID not defined.");
-	} else {
-		console.log("BUILD_ID", process.env.BUILD_ID);
+		throw new Error("BUILD_ID environment variable is not set.");
 	}
+	if (process.env.PIPELINE === undefined) {
+		throw new Error("PIPELINE environment variable is not set.");
+	}
+
+	console.log("BUILD_ID:", process.env.BUILD_ID);
+	console.log("PIPELINE:", process.env.PIPELINE);
+	console.log("STAGE_ID:", process.env.STAGE_ID);
 
-	// Note: type == "Task" would include tasks from the stages in the result set. It might be interesting in the future - for now we will only collect stages.
 	const parsedJobs: ParsedJob[] = fileData.records
-		.filter((job) => job.type === "Stage")
-		.map((job) => {
-			const startTime = Date.parse(job.startTime?.toString()) ?? undefined;
-			const finishTime = Date.parse(job.finishTime?.toString()) ?? undefined;
-			const dateDiff =
-				finishTime && startTime ? Math.abs(finishTime - startTime) / 1000 : undefined; // diff in seconds
-			console.log(`Name=${job.name}`);
+		// Note: type === "Task" or type === "Job" would include task-level (or job-level, respectively) telemetry.
+		// It might be interesting in the future - for now we will only collect stage-level telemetry.
+		// If given a specific STAGE_ID, only process that stage. Otherwise process all stages.
+		.filter((job) => job.type === "Stage" && (process.env.STAGE_ID === undefined || job.identifier === process.env.STAGE_ID))
+		.map((job): ParsedJob | undefined => {
+			console.log(
+				`Processing stage - name='${job.name}' identifier='${job.identifier}' state='${job.state}' result='${job.result}'`,
+			);
+
+			const finishTime = Date.parse(job.finishTime?.toString());
+			if (Number.isNaN(finishTime)) {
+				console.error(
+					`Failed to parse finishTime '${job.finishTime}'. The specified pipeline stage might not have finished yet. Telemetry for this stage will not be sent.`,
+				);
+				return undefined;
+			}
+
+			let startTime: number = finishTime;
+			if (job.state === "completed" && job.startTime === null) {
+				// A null start time when 'state === completed' indicates the stage was skipped.
+				// Set startTime to finishTime so duration ends up being 0.
+				startTime = finishTime;
+			} else {
+				startTime = Date.parse(job.startTime?.toString());
+				if (Number.isNaN(startTime)) {
+					console.error(`Failed to parse startTime '${job.startTime}'. Telemetry for this stage will not be sent.`);
+					return undefined;
+				}
+			}
+
 			return {
 				// Using the 'identifier' property because that's the one available in the API response for test results,
 				// and we want the values to be consistent so we can correlate them later.
 				stageName: job.identifier,
 				startTime,
 				finishTime,
-				totalTime: dateDiff,
+				totalSeconds: Math.abs(finishTime - startTime) / 1000,
 				state: job.state,
```

Comment:
Please comment why `Math.abs` is needed here

Reply:
Probably not necessary, I just moved existing code around but didn't think too hard to optimize it. I'll remove it here, since the new checks above should take care of potential problems with one of the two not existing.